inference/conflict.go

```diff
@@ -185,11 +213,38 @@ func (l *conflictList) addOverconstraintConflict(nilExplanation ExplainedBool, n
 
 func (l *conflictList) diagnostics() []analysis.Diagnostic {
 	var diagnostics []analysis.Diagnostic
-	for _, c := range l.conflicts {
+
+	// group conflicts with the same nil path together for concise reporting
+	groupedConflicts := groupConflicts(l.conflicts)
+
+	// build diagnostics from grouped conflicts
+	for _, c := range groupedConflicts {
 		diagnostics = append(diagnostics, analysis.Diagnostic{
 			Pos:     c.position,
 			Message: c.String(),
 		})
 	}
 	return diagnostics
 }
+
+// groupConflicts groups conflicts with the same nil path together and update conflicts list.
+func groupConflicts(allConflicts []conflict) []conflict {
+	conflictsMap := make(map[string]conflict)
+	for _, c := range allConflicts {
+		s := pathString(c.nilFlow.nilPath)
+		if existingConflict, ok := conflictsMap[s]; ok {
+			// Grouping condition satisfied. Add new conflict to `similarConflicts` in `existingConflict`, and update groupedConflicts map
+			existingConflict.addSimilarConflict(c)
+			conflictsMap[s] = existingConflict
+		} else {
+			conflictsMap[s] = c
+		}
+	}
+
+	// update groupedConflicts list with grouped groupedConflicts
+	var groupedConflicts []conflict
+	for _, c := range conflictsMap {
```

Comment:
Use of map iteration here will lead to non-deterministic printing of errors. Can we iterate over `allConflicts` instead, skipping those not in `conflictsMap`? Or could we even populate this list in the for loop above on the else branch? Might need switching some stuff to pointer types :)

Reply:
Good point. Done :)